log-warper/src/Log/Monad.hs

```diff
@@ -0,0 +1,120 @@
+{-# LANGUAGE ConstraintKinds        #-}
+{-# LANGUAGE DataKinds              #-}
+{-# LANGUAGE DefaultSignatures      #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE TemplateHaskell        #-}
+{-# LANGUAGE UndecidableInstances   #-}
+
+-- | Contains base monad for all logging actions in @log-warper@ library. This
+-- monad keeps logging 'Context' inside 'ReaderT'.
+
+module Log.Monad
+       ( -- * Logger Context
+         Context (..)
+       , ctxName
+       , ctxConfiguration
+
+         -- * Logger monads
+       , LoggerT (..)
+       , MonadLogContext (..)
+       , WarperAction
+       , WithLogger
+       , launchLogger
+
+         -- * Logging functions
+       , logM
+       , withSublogger
+       ) where
+
+import Universum
+
+import Lens.Micro.Platform (makeLenses)
+
+import Log.Configuration (Configuration, Extension)
+import Log.Core.Action (LogAction (..))
+import Log.Core.Class (ActionT, MonadLogger, logR, usingActionT, withR)
+import Log.Event (Event (..), eName)
+import Log.Name (LoggerName)
+import Log.Severity (Severity)
+
+-- | Runtime context for main 'LoggerT' monad. Keeps current 'LoggerName' inside
+-- 'ReaderT' context as well as global configuration.
+data Context (exts :: [Extension]) = Context
+    { _ctxName          :: !LoggerName
+    , _ctxConfiguration :: !(Configuration exts)
+    }
+
+makeLenses ''Context
+
+-- | Logging monad with access to 'Context'.
+newtype LoggerT exts m a = LoggerT
+    { runLoggerT :: ReaderT (Context exts) m a
+    } deriving (Functor, Applicative, Monad, MonadIO, MonadTrans)
+
+-- | Type for all logging actions which are supposed to be used in @log-warper@.
+type WarperAction exts m = LogAction (LoggerT exts m) (Event exts)
+
+{- | Type class for monads which have access to 'Context'.
+
+__Laws:__
+
+* __@Idempotency:@__ context is not changed over time
+
+@
+askLogContext >> askLogContext ≡ askLogContext
```

Comment:
This law does not assure that the context is not changed over time. It only assures that `askLogContext` doesn't change itself.

I can say that for `StateT` we have:

```
get >> get = get
```

And yet there's `put`.